### PR TITLE
[#IOPAE-247] Service Publication Watcher

### PR DESCRIPTION
**Pull request by giuseppedipinto**
Added _ServicePublicationWatcher_ that listen changes on _services-pubblications_ cosmosdb container and produce a _requestHistory_ event on _request-history_ queue.
Then, an _OnRequestHistory_ function listen to _requestHistory_ queue item and write it on _services-history_ cosmosdb container.

**Note:**
_OnRequestHistory_ function use _Azure Cosmos DB output binding_ for write `requestHistoryItem` _(heard from the input queue)_ to database _(cosmosDB output binding)_.
#### List of Changes
- Added _**request-history**_ queue on `infra`
- Added _**ServicePublicationWatcher**_ function _(input: cosmosDBTrigger, output: request-history queue item)_
- Added _**OnRequestHistory**_ function _(input: queueTrigger, output: write on db)_
- Added unit tests

#### How Has This Been Tested?
Unit tests

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/io-services-cms-webapp/ServicePublicationWatcher/function.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      "collectionName": "%COSMOSDB_CONTAINER_SERVICES_PUBLICATION%",
+      "leaseCollectionName": "%COSMOSDB_CONTAINER_SERVICES_PUBLICATION%-lease",
+      "createLeaseCollectionIfNotExists": true,
+      "startFromBeginning": true,
+      "maxItemsPerInvocation": 30
+    },
+    {
+      "queueName": "%REQUEST_HISTORY_QUEUE%",
+      "connection": "INTERNAL_STORAGE_CONNECTION_STRING",
+      "name": "requestHistory",
```
Comment (rodoviti) [inline apps/io-services-cms-webapp/ServicePublicationWatcher/function.json:19]:
```suggestion
      "queueName": "%UPDATE_HISTORY_QUEUE%",
      "connection": "INTERNAL_STORAGE_CONNECTION_STRING",
      "name": "updateHistory",
```

**Review by StefanoDoc — APPROVED** [review]

File: apps/io-services-cms-webapp/OnRequestHistory/function.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      "name": "requestHistoryItem",
+      "type": "queueTrigger",
+      "direction": "in"
+    },
+    {
+      "name": "serviceHistoryDocument",
+      "type": "cosmosDB",
+      "databaseName": "%COSMOSDB_NAME%",
+      "collectionName": "%COSMOSDB_CONTAINER_SERVICES_HISTORY%",
+      "createIfNotExists": true,
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/OnRequestHistory/function.json]: [d6b47f6](https://github.com/pagopa/io-services-cms/pull/87/commits/d6b47f6cfa614963a09a469a0e1ff49f8b44ecfa)

File: packages/io-services-cms-models/src/service-lifecycle/definitions.ts [hunk 3]
```diff
@@ -66,10 +66,16 @@ export type ServiceId = t.TypeOf<typeof ServiceId>;
...
+  t.type({
+    id: ServiceId,
+    data: t.intersection([
+      ServiceData,
+      t.type({ organization: OrganizationData, metadata: ServiceMetadata }),
+    ]),
+  }),
+  t.partial({
+    version: NonEmptyString,
+    last_update: NonEmptyString,
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply packages/io-services-cms-models/src/service-lifecycle/definitions.ts]: [d6b47f6](https://github.com/pagopa/io-services-cms/pull/87/commits/d6b47f6cfa614963a09a469a0e1ff49f8b44ecfa)

File: packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts [hunk 4]
```diff
@@ -61,7 +66,14 @@ export const createCosmosStore = <
             }`
           )
       ),
-      TE.map((_) => value)
+      TE.map((itemResponse) => ({
+        ...value,
+        last_update: itemResponse.resource
+          ? // eslint-disable-next-line no-underscore-dangle
+            new Date(itemResponse.resource._ts * 1000).toISOString() // Unix timestamp
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts]: [d6b47f6](https://github.com/pagopa/io-services-cms/pull/87/commits/d6b47f6cfa614963a09a469a0e1ff49f8b44ecfa)

File: packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts [hunk 5]
```diff
@@ -34,7 +34,12 @@ export const createCosmosStore = <
             TE.right(O.none)
           : // if present, try to decode in the expected shape
             pipe(
-              rr.resource,
+              {
+                ...rr.resource,
+                // eslint-disable-next-line no-underscore-dangle
+                last_update: new Date(rr.resource._ts * 1000).toISOString(), // Unix timestamp
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply packages/io-services-cms-models/src/lib/fsm/store.cosmos.ts]: d6b47f6

File: packages/io-services-cms-models/src/queue/index.ts [hunk 6]
```diff
@@ -6,3 +6,6 @@ export const RequestReviewItem = Service;
 
 export type RequestPublicationItem = t.TypeOf<typeof RequestPublicationItem>;
 export const RequestPublicationItem = Service;
+
+export type RequestHistoryItem = t.TypeOf<typeof RequestHistoryItem>;
+export const RequestHistoryItem = Service;
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply packages/io-services-cms-models/src/queue/index.ts]: https://github.com/pagopa/io-services-cms/commit/d6b47f6cfa614963a09a469a0e1ff49f8b44ecfa

File: apps/io-services-cms-webapp/src/historicizer/request-history-handler.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import { IsoDateFromString } from "@pagopa/ts-commons/lib/dates";
+import { withJsonInput } from "../lib/azure/misc";
+
+const parseIncomingMessage = (
+  queueItem: Json
+): E.Either<Error, ServiceLifecycle.definitions.Service> =>
+  pipe(
+    queueItem,
+    ServiceLifecycle.definitions.Service.decode,
+    E.mapLeft(flow(readableReport, (_) => new Error(_)))
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/src/historicizer/request-history-handler.ts]: https://github.com/pagopa/io-services-cms/commit/d6b47f6cfa614963a09a469a0e1ff49f8b44ecfa

File: apps/io-services-cms-webapp/env.example [hunk 8]
```diff
@@ -46,3 +46,4 @@ COSMOSDB_CONTAINER_SERVICES_HISTORY=
 # Queue configuration
 REQUEST_REVIEW_QUEUE=
 REQUEST_PUBLICATION_QUEUE=
+REQUEST_HISTORY_QUEUE=
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/io-services-cms-webapp/env.example]: 1af46fd

**Review by StefanoDoc — APPROVED** [review]

**Pull request merged**